File: sdk-swift/Sources/SlyOS/AudioProcessor.swift
import Foundation
import AVFoundation
import Accelerate

// MARK: - Audio Processor for Whisper

/// Handles audio loading and mel spectrogram computation for Whisper speech-to-text models.
/// Uses AVFoundation for audio decoding and Accelerate for DSP operations.
public final class AudioProcessor {

    /// Whisper expects audio at 16kHz sample rate
    public static let sampleRate: Float = 16000.0

    /// Number of mel filterbank channels (Whisper uses 80)
    public static let nMels: Int = 80

    /// FFT size for STFT
    public static let nFFT: Int = 400

    /// Hop length between frames
    public static let hopLength: Int = 160

    /// Maximum audio length in samples (30 seconds at 16kHz)
    public static let maxAudioLength: Int = 480000

    // MARK: - Audio Loading

    /// Load an audio file and convert to 16kHz mono float samples.
    /// - Parameter url: URL to the audio file (supports wav, mp3, m4a, caf, etc.)
    /// - Returns: Array of float samples normalized to [-1, 1]
    public static func loadAudio(from url: URL) throws -> [Float] {
        guard FileManager.default.fileExists(atPath: url.path) else {
            throw AudioProcessorError.fileNotFound(url.path)
        }

        let audioFile = try AVAudioFile(forReading: url)

        // Set up format for 16kHz mono Float32
        guard let targetFormat = AVAudioFormat(
            commonFormat: .pcmFormatFloat32,
            sampleRate: Double(sampleRate),
            channels: 1,
            interleaved: false
        ) else {
            throw AudioProcessorError.formatError("Cannot create target audio format")
        }

        // Create converter
        guard let converter = AVAudioConverter(from: audioFile.processingFormat, to: targetFormat) else {
            throw AudioProcessorError.formatError("Cannot create audio converter from \(audioFile.processingFormat) to \(targetFormat)")
        }

        // Calculate output frame count
        let ratio = Double(sampleRate) / audioFile.fileFormat.sampleRate
        let outputFrameCount = AVAudioFrameCount(Double(audioFile.length) * ratio)

        guard let outputBuffer = AVAudioPCMBuffer(pcmFormat: targetFormat, frameCapacity: outputFrameCount) else {
            throw AudioProcessorError.formatError("Cannot create output buffer")
        }

        // Read and convert
        let inputBuffer = AVAudioPCMBuffer(pcmFormat: audioFile.processingFormat, frameCapacity: AVAudioFrameCount(audioFile.length))!
        try audioFile.read(into: inputBuffer)

        var error: NSError?
        converter.convert(to: outputBuffer, error: &error) { _, outStatus in
            outStatus.pointee = .haveData
            return inputBuffer
        }

        if let error = error {
            throw AudioProcessorError.conversionError(error.localizedDescription)
        }

        // Extract float samples
        guard let channelData = outputBuffer.floatChannelData else {
            throw AudioProcessorError.formatError("No channel data in output buffer")
        }

        let frameCount = Int(outputBuffer.frameLength)
        let samples = Array(UnsafeBufferPointer(start: channelData[0], count: frameCount))

        return samples
    }

    // MARK: - Mel Spectrogram

    /// Compute log mel spectrogram from audio samples, matching Whisper's preprocessing.
    /// - Parameter audio: Audio samples at 16kHz, normalized to [-1, 1]
    /// - Returns: 2D array of shape [n_mels, n_frames]
    public static func computeMelSpectrogram(audio: [Float]) -> [[Float]] {
        // Pad or truncate to max length
        var paddedAudio: [Float]
        if audio.count > maxAudioLength {
            paddedAudio = Array(audio.prefix(maxAudioLength))
        } else {
            paddedAudio = audio
            if audio.count < maxAudioLength {
                paddedAudio.append(contentsOf: Array(repeating: 0, count: maxAudioLength - audio.count))
            }
        }

        // Compute STFT
        let nFrames = (paddedAudio.count - nFFT) / hopLength + 1
        var magnitudes = [[Float]](repeating: [Float](repeating: 0, count: nFFT / 2 + 1), count: nFrames)

        // Hann window
        let window = hannWindow(size: nFFT)

        // Process each frame
        for frame in 0..<nFrames {
            let start = frame * hopLength
            let end = min(start + nFFT, paddedAudio.count)

            // Extract and window the frame
            var windowed = [Float](repeating: 0, count: nFFT)
            let frameLength = end - start
            for i in 0..<frameLength {
                windowed[i] = paddedAudio[start + i] * window[i]
            }

            // Compute magnitude spectrum using Accelerate vDSP FFT
            let spectrum = fftMagnitude(signal: windowed)
            magnitudes[frame] = spectrum
        }

        // Generate mel filterbank
        let melFilters = melFilterbank(
            nMels: nMels,
            nFFT: nFFT,
            sampleRate: sampleRate
        )

        // Apply mel filterbank: [n_mels x (nFFT/2+1)] * [n_frames x (nFFT/2+1)]^T
        var melSpec = [[Float]](repeating: [Float](repeating: 0, count: nFrames), count: nMels)

        for mel in 0..<nMels {
            for frame in 0..<nFrames {
                var sum: Float = 0
                for bin in 0..<(nFFT / 2 + 1) {
                    sum += melFilters[mel][bin] * magnitudes[frame][bin]
                }
                melSpec[mel][frame] = sum
            }
        }

        // Log mel spectrogram
        let logOffset: Float = 1e-10
        for mel in 0..<nMels {
            for frame in 0..<nFrames {
                melSpec[mel][frame] = log10(max(melSpec[mel][frame], logOffset))
            }
        }

        // Normalize: clamp to max - 8.0, then scale to [0, 1] * 4.0
        let maxVal = melSpec.flatMap { $0 }.max() ?? 0
        let minClamp = maxVal - 8.0
        for mel in 0..<nMels {
            for frame in 0..<nFrames {
                melSpec[mel][frame] = max(melSpec[mel][frame], minClamp)
                melSpec[mel][frame] = (melSpec[mel][frame] - minClamp) / 8.0 * 4.0
            }
        }

        return melSpec
    }

    // MARK: - DSP Helpers

    /// Create a Hann window of given size.
    private static func hannWindow(size: Int) -> [Float] {
        var window = [Float](repeating: 0, count: size)
        vDSP_hann_window(&window, vDSP_Length(size), Int32(vDSP_HANN_NORM))
        return window
    }

    /// Compute magnitude spectrum of a signal using vDSP FFT.
    private static func fftMagnitude(signal: [Float]) -> [Float] {
        let n = signal.count
        let log2n = vDSP_Length(log2(Float(n)))

        guard let fftSetup = vDSP_create_fftsetup(log2n, FFTRadix(kFFTRadix2)) else {
            return [Float](repeating: 0, count: n / 2 + 1)
        }
        defer { vDSP_destroy_fftsetup(fftSetup) }

        // Split complex arrays
        var realPart = [Float](repeating: 0, count: n / 2)
        var imagPart = [Float](repeating: 0, count: n / 2)

        // Convert to split complex format
        signal.withUnsafeBufferPointer { signalPtr in
            signalPtr.baseAddress!.withMemoryRebound(to: DSPComplex.self, capacity: n / 2) { complexPtr in
                var splitComplex = DSPSplitComplex(realp: &realPart, imagp: &imagPart)
                vDSP_ctoz(complexPtr, 2, &splitComplex, 1, vDSP_Length(n / 2))
            }
        }

        // Perform FFT
        var splitComplex = DSPSplitComplex(realp: &realPart, imagp: &imagPart)
        vDSP_fft_zrip(fftSetup, &splitComplex, 1, log2n, FFTDirection(kFFTDirection_Forward))

        // Compute magnitude squared
        var magnitudes = [Float](repeating: 0, count: n / 2 + 1)

        // DC component
        magnitudes[0] = realPart[0] * realPart[0]
        // Nyquist component
        magnitudes[n / 2] = imagPart[0] * imagPart[0]

        // Other components
        for i in 1..<(n / 2) {
            magnitudes[i] = realPart[i] * realPart[i] + imagPart[i] * imagPart[i]
        }

        // Scale
        var scale: Float = 1.0 / Float(n * n)
        vDSP_vsmul(magnitudes, 1, &scale, &magnitudes, 1, vDSP_Length(magnitudes.count))

        return magnitudes
    }

    /// Generate mel filterbank matrix.
    private static func melFilterbank(nMels: Int, nFFT: Int, sampleRate: Float) -> [[Float]] {
        let nBins = nFFT / 2 + 1

        // Mel scale conversion
        func hzToMel(_ hz: Float) -> Float {
            return 2595.0 * log10(1.0 + hz / 700.0)
        }

        func melToHz(_ mel: Float) -> Float {
            return 700.0 * (pow(10.0, mel / 2595.0) - 1.0)
        }

        let melMin = hzToMel(0)
        let melMax = hzToMel(sampleRate / 2)

        // Create equally-spaced mel points
        var melPoints = [Float](repeating: 0, count: nMels + 2)
        for i in 0...(nMels + 1) {
            melPoints[i] = melMin + Float(i) * (melMax - melMin) / Float(nMels + 1)
        }

        // Convert mel points to frequency bins
        let fftFreqs = (0..<nBins).map { Float($0) * sampleRate / Float(nFFT) }
        let melFreqs = melPoints.map { melToHz($0) }

        // Build filterbank
        var filterbank = [[Float]](repeating: [Float](repeating: 0, count: nBins), count: nMels)

        for mel in 0..<nMels {
            let lower = melFreqs[mel]
            let center = melFreqs[mel + 1]
            let upper = melFreqs[mel + 2]

            for bin in 0..<nBins {
                let freq = fftFreqs[bin]

                if freq >= lower && freq <= center && center > lower {
                    filterbank[mel][bin] = (freq - lower) / (center - lower)
                } else if freq > center && freq <= upper && upper > center {
                    filterbank[mel][bin] = (upper - freq) / (upper - center)
                }
            }
        }

        return filterbank
    }
}

// MARK: - Audio Processor Errors

public enum AudioProcessorError: LocalizedError {
    case fileNotFound(String)
    case formatError(String)
    case conversionError(String)
    case processingError(String)

    public var errorDescription: String? {
        switch self {
        case .fileNotFound(let path):
            return "Audio file not found: \(path)"
        case .formatError(let msg):
            return "Audio format error: \(msg)"
        case .conversionError(let msg):
            return "Audio conversion error: \(msg)"
        case .processingError(let msg):
            return "Audio processing error: \(msg)"
        }
    }
}

File: sdk-kotlin/src/main/kotlin/com/slyos/AudioProcessor.kt
package com.slyos

import android.content.Context
import android.media.MediaCodec
import android.media.MediaExtractor
import android.media.MediaFormat
import android.net.Uri
import java.io.File
import java.nio.ByteBuffer
import java.nio.ByteOrder
import kotlin.math.*

/**
 * Handles audio loading and mel spectrogram computation for Whisper speech-to-text models.
 * Uses Android MediaCodec for audio decoding and custom FFT for mel spectrogram.
 */
object AudioProcessor {

    /** Whisper expects audio at 16kHz */
    const val SAMPLE_RATE = 16000f

    /** Number of mel filterbank channels (Whisper uses 80) */
    const val N_MELS = 80

    /** FFT size for STFT */
    const val N_FFT = 400

    /** Hop length between frames */
    const val HOP_LENGTH = 160

    /** Maximum audio length in samples (30 seconds at 16kHz) */
    const val MAX_AUDIO_LENGTH = 480000

    // ──────────────────────────────────────────────────────────────────

    /**
     * Load an audio file and convert to 16kHz mono float samples.
     *
     * @param context Android context
     * @param audioUri URI to the audio file
     * @return Float array of audio samples normalized to [-1, 1]
     */
    fun loadAudio(context: Context, audioUri: Uri): FloatArray {
        val extractor = MediaExtractor()

        // Set data source
        val fd = context.contentResolver.openFileDescriptor(audioUri, "r")
            ?: throw AudioProcessingException("Cannot open audio file: $audioUri")

        extractor.setDataSource(fd.fileDescriptor)
        fd.close()

        // Find audio track
        var audioTrackIndex = -1
        var audioFormat: MediaFormat? = null
        for (i in 0 until extractor.trackCount) {
            val format = extractor.getTrackFormat(i)
            val mime = format.getString(MediaFormat.KEY_MIME) ?: continue
            if (mime.startsWith("audio/")) {
                audioTrackIndex = i
                audioFormat = format
                break
            }
        }

        if (audioTrackIndex == -1 || audioFormat == null) {
            extractor.release()
            throw AudioProcessingException("No audio track found in file")
        }

        extractor.selectTrack(audioTrackIndex)

        val sourceSampleRate = audioFormat.getInteger(MediaFormat.KEY_SAMPLE_RATE)
        val channelCount = audioFormat.getInteger(MediaFormat.KEY_CHANNEL_COUNT)
        val mime = audioFormat.getString(MediaFormat.KEY_MIME) ?: "audio/raw"

        // Decode audio using MediaCodec
        val decoder = MediaCodec.createDecoderByType(mime)
        decoder.configure(audioFormat, null, null, 0)
        decoder.start()

        val samples = mutableListOf<Float>()
        val bufferInfo = MediaCodec.BufferInfo()
        var isEOS = false

        while (!isEOS) {
            // Feed input
            val inputIndex = decoder.dequeueInputBuffer(10000)
            if (inputIndex >= 0) {
                val inputBuffer = decoder.getInputBuffer(inputIndex) ?: continue
                val sampleSize = extractor.readSampleData(inputBuffer, 0)

                if (sampleSize < 0) {
                    decoder.queueInputBuffer(inputIndex, 0, 0, 0, MediaCodec.BUFFER_FLAG_END_OF_STREAM)
                    isEOS = true
                } else {
                    val presentationTime = extractor.sampleTime
                    decoder.queueInputBuffer(inputIndex, 0, sampleSize, presentationTime, 0)
                    extractor.advance()
                }
            }

            // Read output
            var outputIndex = decoder.dequeueOutputBuffer(bufferInfo, 10000)
            while (outputIndex >= 0) {
                val outputBuffer = decoder.getOutputBuffer(outputIndex) ?: break
                outputBuffer.order(ByteOrder.nativeOrder())

                // Decode PCM 16-bit samples
                val shortBuffer = outputBuffer.asShortBuffer()
                val shortArray = ShortArray(shortBuffer.remaining())
                shortBuffer.get(shortArray)

                // Convert to mono float [-1, 1]
                for (i in shortArray.indices step channelCount) {
                    val sample = shortArray[i].toFloat() / Short.MAX_VALUE
                    samples.add(sample)
                }

                decoder.releaseOutputBuffer(outputIndex, false)
                outputIndex = decoder.dequeueOutputBuffer(bufferInfo, 0)
            }
        }

        decoder.stop()
        decoder.release()
        extractor.release()

        // Resample to 16kHz if needed
        val result = if (sourceSampleRate != SAMPLE_RATE.toInt()) {
            resample(samples.toFloatArray(), sourceSampleRate, SAMPLE_RATE.toInt())
        } else {
            samples.toFloatArray()
        }

        return result
    }

    /**
     * Load audio from a file path.
     */
    fun loadAudio(context: Context, filePath: String): FloatArray {
        return loadAudio(context, Uri.fromFile(File(filePath)))
    }

    // ──────────────────────────────────────────────────────────────────

    /**
     * Compute log mel spectrogram from audio samples, matching Whisper's preprocessing.
     *
     * @param audio Audio samples at 16kHz, normalized to [-1, 1]
     * @return 2D array of shape [N_MELS x n_frames]
     */
    fun computeMelSpectrogram(audio: FloatArray): Array<FloatArray> {
        // Pad or truncate to max length
        val paddedAudio = when {
            audio.size > MAX_AUDIO_LENGTH -> audio.copyOfRange(0, MAX_AUDIO_LENGTH)
            audio.size < MAX_AUDIO_LENGTH -> audio + FloatArray(MAX_AUDIO_LENGTH - audio.size)
            else -> audio
        }

        // Compute STFT
        val nFrames = (paddedAudio.size - N_FFT) / HOP_LENGTH + 1
        val magnitudes = Array(nFrames) { FloatArray(N_FFT / 2 + 1) }

        // Hann window
        val window = hannWindow(N_FFT)

        // Process each frame
        for (frame in 0 until nFrames) {
            val start = frame * HOP_LENGTH
            val windowed = FloatArray(N_FFT)
            for (i in 0 until minOf(N_FFT, paddedAudio.size - start)) {
                windowed[i] = paddedAudio[start + i] * window[i]
            }

            magnitudes[frame] = fftMagnitude(windowed)
        }

        // Generate mel filterbank
        val melFilters = melFilterbank(N_MELS, N_FFT, SAMPLE_RATE)

        // Apply mel filterbank
        val melSpec = Array(N_MELS) { mel ->
            FloatArray(nFrames) { frame ->
                var sum = 0f
                for (bin in 0 until N_FFT / 2 + 1) {
                    sum += melFilters[mel][bin] * magnitudes[frame][bin]
                }
                sum
            }
        }

        // Log mel spectrogram with normalization
        val logOffset = 1e-10f
        var maxVal = Float.NEGATIVE_INFINITY
        for (mel in 0 until N_MELS) {
            for (frame in 0 until nFrames) {
                melSpec[mel][frame] = log10(maxOf(melSpec[mel][frame], logOffset))
                if (melSpec[mel][frame] > maxVal) maxVal = melSpec[mel][frame]
            }
        }

        // Normalize: clamp and scale
        val minClamp = maxVal - 8.0f
        for (mel in 0 until N_MELS) {
            for (frame in 0 until nFrames) {
                melSpec[mel][frame] = maxOf(melSpec[mel][frame], minClamp)
                melSpec[mel][frame] = (melSpec[mel][frame] - minClamp) / 8.0f * 4.0f
            }
        }

        return melSpec
    }

    // ──────────────────────────────────────────────────────────────────

    /** Simple linear resampling. */
    private fun resample(audio: FloatArray, fromRate: Int, toRate: Int): FloatArray {
        val ratio = toRate.toDouble() / fromRate
        val outputLength = (audio.size * ratio).toInt()
        val output = FloatArray(outputLength)

        for (i in output.indices) {
            val srcIndex = i / ratio
            val srcInt = srcIndex.toInt()
            val frac = (srcIndex - srcInt).toFloat()

            output[i] = if (srcInt + 1 < audio.size) {
                audio[srcInt] * (1 - frac) + audio[srcInt + 1] * frac
            } else {
                audio[minOf(srcInt, audio.size - 1)]
            }
        }
        return output
    }

    /** Hann window. */
    private fun hannWindow(size: Int): FloatArray {
        return FloatArray(size) { i ->
            0.5f * (1 - cos(2.0 * PI * i / size).toFloat())
        }
    }

    /** Compute magnitude spectrum via naive DFT (for frames of size N_FFT=400). */
    private fun fftMagnitude(signal: FloatArray): FloatArray {
        val n = signal.size
        val nBins = n / 2 + 1
        val magnitudes = FloatArray(nBins)

        // Naive DFT — O(n²) but n is small (400) so ~160K ops per frame, acceptable
        for (k in 0 until nBins) {
            var real = 0f
            var imag = 0f
            val omega = -2.0 * PI * k / n
            for (t in signal.indices) {
                val angle = (omega * t).toFloat()
                real += signal[t] * cos(angle)
                imag += signal[t] * sin(angle)
            }
            magnitudes[k] = (real * real + imag * imag) / (n * n).toFloat()
        }

        return magnitudes
    }

    /** Generate mel filterbank matrix. */
    private fun melFilterbank(nMels: Int, nFFT: Int, sampleRate: Float): Array<FloatArray> {
        val nBins = nFFT / 2 + 1

        fun hzToMel(hz: Float) = 2595.0f * log10(1.0f + hz / 700.0f)
        fun melToHz(mel: Float) = 700.0f * (10.0f.pow(mel / 2595.0f) - 1.0f)

        val melMin = hzToMel(0f)
        val melMax = hzToMel(sampleRate / 2)

        val melPoints = FloatArray(nMels + 2) { i ->
            melMin + i * (melMax - melMin) / (nMels + 1)
        }

        val fftFreqs = FloatArray(nBins) { it * sampleRate / nFFT }
        val melFreqs = melPoints.map { melToHz(it) }

        return Array(nMels) { mel ->
            FloatArray(nBins) { bin ->
                val freq = fftFreqs[bin]
                val lower = melFreqs[mel]
                val center = melFreqs[mel + 1]
                val upper = melFreqs[mel + 2]

                when {
                    freq in lower..center && center > lower -> (freq - lower) / (center - lower)
                    freq in center..upper && upper > center -> (upper - freq) / (upper - center)
                    else -> 0f
                }
            }
        }
    }
}

/** Exception thrown during audio processing. */
class AudioProcessingException(message: String) : Exception(message)
